app/experimenter/nimbus-ui/src/components/FormBranches/reducer.tsx

```diff
@@ -0,0 +1,266 @@
+  } else {
+    treatmentBranches = [
+      ...(treatmentBranches || []),
+      makeBranch(state.lastId, `New treatment ${lastId}`),
+    ];
+  }
+
+  return {
+    ...state,
+    lastId,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+type RemoveBranchAction = {
+  type: "removeBranch";
+  idx: number;
+};
+
+function removeBranch(state: FormBranchesState, { idx }: RemoveBranchAction) {
+  const treatmentBranches = state.treatmentBranches || [];
+  return {
+    ...state,
+    treatmentBranches: [
+      ...treatmentBranches.slice(0, idx),
+      ...treatmentBranches.slice(idx + 1),
+    ],
+  };
+}
+
+type UpdateBranchAction = {
+  type: "updateBranch";
+  idx: number;
+  value: AnnotatedBranch;
+};
+
+function updateBranch(
+  state: FormBranchesState,
+  { idx, value }: UpdateBranchAction,
+) {
+  let { referenceBranch, treatmentBranches } = state;
+  const updatedBranch = { ...value };
+  if (idx === REFERENCE_BRANCH_IDX) {
+    referenceBranch = updatedBranch;
+  } else {
+    treatmentBranches = (treatmentBranches || []).slice();
+    treatmentBranches[idx] = updatedBranch;
+  }
+  return {
+    ...state,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+type RemoveFeatureConfigAction = {
+  type: "removeFeatureConfig";
+};
+
+function removeFeatureConfig(state: FormBranchesState) {
+  return {
+    ...state,
+    featureConfig: null,
+  };
+}
+
+type SetFeatureConfigAction = {
+  type: "setFeatureConfig";
+  value: FormBranchesState["featureConfig"];
+};
+
+function setFeatureConfig(
+  state: FormBranchesState,
+  { value: featureConfig }: SetFeatureConfigAction,
+) {
+  return {
+    ...state,
+    featureConfig,
+  };
+}
+
+type SetEqualRatioAction = {
+  type: "setEqualRatio";
+  value: FormBranchesState["equalRatio"];
+};
+
+function setEqualRatio(
+  state: FormBranchesState,
+  { value: equalRatio }: SetEqualRatioAction,
+) {
+  let { referenceBranch, treatmentBranches } = state;
+  if (equalRatio) {
+    if (referenceBranch !== null) {
+      referenceBranch = { ...referenceBranch, ratio: 1 };
+    }
+    if (Array.isArray(treatmentBranches)) {
+      treatmentBranches = treatmentBranches.map(
+        (treatmentBranch) =>
+          treatmentBranch && {
+            ...treatmentBranch,
+            ratio: 1,
+          },
+      );
+    }
+  }
+  return {
+    ...state,
+    equalRatio,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+export type AnnotatedBranch = getExperiment_experimentBySlug_treatmentBranches & {
+  __key: string;
+};
+
+function annotateBranch(
+  lastId: number | string,
+  branch: getExperiment_experimentBySlug_treatmentBranches,
+) {
+  return {
+    ...branch,
+    __key: `branch-${lastId}`,
+  };
+}
+
+export function stripOneAnnotatedBranch(branch: null | AnnotatedBranch) {
+  if (!branch) return null;
+  const {
+    __key, // eslint-disable-line @typescript-eslint/no-unused-vars
+    ...strippedBranch
+  } = branch;
+  return strippedBranch;
+}
```

Comment:
~This function has me scratching my head a little. What's the destructuring doing?~ Ooooh it's spreading the rest of the properties to a new object. Clever!

Reply:
Yeah, basically a way to use destructuring to omit certain properties from a new object.